src/datamodel_code_generator/model/template/pydantic_v2/BaseModel.jinja2

```diff
@@ -32,7 +32,7 @@ class {{ class_name }}({{ base_class }}):{% if comment is defined %}  # {{ comme
     {%- else %}
     {{ field.name }}: {{ field.type_hint }}
     {%- endif %}
-    {%- if not (field.required or (field.represented_default == 'None' and field.strip_default_none)) or field.data_type.is_optional
+    {%- if not field.required and (field.represented_default != 'None' or not field.strip_default_none) or field.data_type.is_optional)
```

Comment:
My fault, I forgot a parenthesis.
Should fix https://github.com/koxudaxi/datamodel-code-generator/pull/2520#issuecomment-3523040447

```suggestion
    {%- if not field.required and (field.represented_default != 'None' or not field.strip_default_none or field.data_type.is_optional)
```